src/Renderer/LayeredMaterial.js

```diff
@@ -177,152 +187,248 @@ LayeredMaterial.prototype.dispose = function() {
-    if (this.Textures[layer][slot] === undefined || this.Textures[layer][slot].image === undefined)
-        this.nbTextures[layer] += 1;
+    // Update slot start texture layer
+    for (var j = layerIndex, mx = this.getColorLayersCount(); j < mx; j++)
+        this.uniforms.paramLayers.value[j].x -= texturesCount;
 
-    // BEWARE: array [] -> size: 0; array [10]="wao" -> size: 11
-    this.Textures[layer][slot] = texture ? texture : emptyTexture;
-    this.pitScale[layer][slot] = pitScale ? pitScale : new THREE.Vector3(0.0, 0.0, 1.0);
-
-    // TEMP
-    if (texture === null) {
-        var l = this.slotToLayerIndex(slot);
-        if (l >= 0) {
-            this.paramLayers[l].z = 0;
+    this.texturesCount[l_COLOR] -= texturesCount;
+
+    this.uniforms.offsetScale_L01.value = this.offsetScale[l_COLOR];
+    this.uniforms.dTextures_01.value = this.Textures[l_COLOR];
+
+};
+
+LayeredMaterial.prototype.setTexturesLayer = function(textures, layerType, layer) {
+    var index = this.indexOfColorLayer(layer);
+    var slotOffset = this.getTextureOffsetByLayerIndex(index);
+    for (var i = 0, max = textures.length; i < max; i++) {
+        if (textures[i]) {
+            if(textures[i].texture !== null) {
+                this.setTexture(textures[i].texture, layerType, i + (slotOffset || 0), textures[i].pitch);
+            } else {
+                this.setLayerVisibility(index,false);
+                break;
+            }
         }
     }
+};
+
+LayeredMaterial.prototype.setTexture = function(texture, layerType, slot, offsetScale) {
 
+    if (this.Textures[layerType][slot] === undefined || this.Textures[layerType][slot].image === undefined) {
+        this.texturesCount[layerType] += 1;
+    }
+
+    // BEWARE: array [] -> size: 0; array [10]="wao" -> size: 11
+    this.Textures[layerType][slot] = texture ? texture : emptyTexture;
+    this.offsetScale[layerType][slot] = offsetScale ? offsetScale : new THREE.Vector3(0.0, 0.0, 1.0);
 };
 
-LayeredMaterial.prototype.slotToLayerIndex = function(slot) {
-    for (var l = 1; l < this.uniforms.nColorLayer.value; l++) {
-        if (slot < this.paramLayers[l].x) {
-            return l-1;
+LayeredMaterial.prototype.setColorLayerParameters = function(params) {
+    if(this.getColorLayersCount() === 0) {
+        for (var l = 0; l < params.length; l++) {
+            this.pushLayer(params[l]);
         }
     }
-    return this.uniforms.nColorLayer.value - 1;
 };
 
-LayeredMaterial.prototype.getNbColorTexturesLayer = function(layerIndex) {
-    return (this.paramLayers[layerIndex + 1].x || this.nbTextures[1]) - (this.paramLayers[layerIndex].x || 0);
+LayeredMaterial.prototype.pushLayer = function(param) {
+
+    let newIndex = this.getColorLayersCount();
+    let offset = newIndex === 0 ? 0 : this.getTextureOffsetByLayerIndex(newIndex-1) + this.getTextureCountByLayerIndex(newIndex-1);
+
+    this.uniforms.paramLayers.value[newIndex] = new THREE.Vector4();
+
+    this.setLayerOffset(newIndex,offset);
```

Comment:
You could rename `setLayerOffset` : `setTextureOffsetByLayerIndex` to match its getter